testutils: Automated testing environment for release tests using pytest

miri64 opened the pull request:
### Contribution description

Third time's a charm: after https://github.com/RIOT-OS/Release-Specs/pull/79 and some major overhaul of that framework to use (then) `riotnode` (now `riotctrl`) and `pytest` by @fjmolinas, I ported the latter to run with the new [`riotctrl`](https://github.com/RIOT-OS/riotctrl/) python package and the in-tree [`riotctrl_shell`](https://github.com/RIOT-OS/RIOT/tree/b05b2fddb967b170f2c2cdcfda5641085d55fbcc/dist/pythonlibs/riotctrl_shell) module.

I also cleaned up the by now very complicated history of that effort and squashed a lot of commits. Since the most of the porting work was done by @fjmolinas (I just moved some code around and enhanced his code) I made him the main author of the work in the commit history, but added all co-authors to the respective commits for the changes they made.

### Testing procedure
Run `tox` and the tests will be run (there are none at the moment, see #160, #161, #162, #163, #158, and #159).
Check-out #164 to run all tests.
There is also some more info in the README how to select specific groups of tests.

Self-tests are run by Github Actions and can be run locally with `tox -- --self-test`.

### Issues/PRs references
Alternative to #79.

Required for #160, #161, #162, #163, #158, and #159

miri64 commented:
Spec06 is still a bit flaky due to the custom `ShellInteraction` for the `udp` command, that's why they are marked as WIP and currently still skipped in the test execution.

miri64 commented:
Fixed the issue with spec 6. One issue was just a wrong assertion not in line with the release-specs and the other was that the timeout for the `udp` command was selected too short.

aabadie commented on testutils/iotlab.py:
```diff
@@ -0,0 +1,171 @@
+        'pba-d-01-kw2x': {'name': 'phynode', 'radio': 'kw2xrf'},
+        'samr21-xpro': {'name': 'samr21', 'radio': 'at86rf233'},
+        'samr30-xpro': {'name': 'samr30', 'radio': 'at86rf212b'},
+        'wsn430-v1_3b': {'name': 'wsn430', 'radio': 'cc1101'},
+        'wsn430-v1_4': {'name': 'wsn430', 'radio': 'cc2420'},
```
The WSN430 were removed this week from iot-lab and will no longer be available (EOL). So I think those 2 lines could be removed (and the boards removed from RIOT as well)

miri64 replied:
:heavy_check_mark:

miri64 commented on conftest.py:
```diff
@@ -0,0 +1,266 @@
+            item.add_marker(iotlab_creds)
+
+
+def pytest_keyboard_interrupt(excinfo):
+    # pylint: disable=C0301
+    """
+    Called on KeyInterrupt
+
+    See: https://docs.pytest.org/en/stable/reference.html?highlight=hooks#_pytest.hookspec.pytest_keyboard_interrupt
+    """     # noqa: E501
+    for child in RUNNING_CTRLS:
+        child.stop_term()
+    for exp in RUNNING_EXPERIMENTS:
+        exp.stop()
```
Another thing that is flaky, even with this function, is cleanup. When the tests run uninterrupted, everything is cleaned up properly, but once one hits `Ctrl+C` during the tests some nodes and experiments might still be left running. Maybe someone with more experience with `pytest` can help here. @aabadie @fjmolinas?

miri64 replied:
(it seems that both commands wait a bit after doing the actual stopping while pytest is too impatient and just kills the waiting).

miri64 commented:
> Fixed the issue with spec 6. One issue was just a wrong assertion not in line with the release-specs and the other was that the timeout for the `udp` command was selected too short.

Mh nope, the timeout is allegedly still too short :-/. This doesn't seem right to me. It is calculated here from the input parameters:

https://github.com/RIOT-OS/Release-Specs/blob/0ddb383f287dd7837bbf4c904dd7263d6cfd735b/testutils/shell.py#L125-L128

So there _should_ be more than enough time to wait for the prompt. However, it does not seem to be and the framework just fires a timeout in

https://github.com/RIOT-OS/Release-Specs/blob/0ddb383f287dd7837bbf4c904dd7263d6cfd735b/testutils/shell.py#L129-L134

Maybe the nodes don't really wait a `delay_us` between sends but much longer? :-/ This problem does not appear with the `ping6` interaction.

miri64 commented:
> Maybe the nodes don't really wait a `delay_us` between sends but much longer?

Yeah seems like this. Instead of waiting 1s with `delay_ms=1000` its more like 1.1s (since the app builds the whole package at every iteration this isn't surprising). Added up to 1000 packets that's 100sec the `replwrap` needs to wait longer for the prompt. I guess I make the timeout slack adaptive to `count` then.

miri64 commented:
> Added up to 1000 packets that's 100sec the `replwrap` needs to wait longer for the prompt. I guess I make the timeout slack adaptive to `count` then.

Yepp, that worked :-)

miri64 commented:
Squashed the last few commits regarding GitHub Actions into one. They are now working https://github.com/miri64/Release-Specs/actions/runs/164755513. For now, only the self-testing of the test framework is done and I plan to keep it that way for pull requests and pushes. If this gets merged I plan to provide a separate workflow to run release tests periodically (maybe weekly?).

miri64 commented:
First run gone horribly wrong: https://miri64.uber.space/test-report-rc1-2020-07-11-12-44.html

miri64 commented:
> https://miri64.uber.space/test-report-rc1-2020-07-11-12-44.html

I used [`junit2html`](https://pypi.org/project/junit2html/) to convert the `test-report.xml` generated by `pytest` to an HTML :blush:

miri64 commented:
With the latest round of commits the success rate should be much higher (also include some clean-up of code I was annoyed with). I'm just running the script again and will come back with the results.

miri64 commented:
> I'm just running the script again and will come back with the results.

See https://github.com/RIOT-OS/Release-Specs/issues/156#issuecomment-657118379. There only seems to be an issue with `asyncio` now.

miri64 commented:
> See [#156 (comment)](https://github.com/RIOT-OS/Release-Specs/issues/156#issuecomment-657118379). There only seems to be an issue with `asyncio` now.

From that comment: does anybody know how to get the `stdout` of failed tests into the report XML?

miri64 commented:
> From that comment: does anybody know how to get the `stdout` of failed tests into the report XML?

[Found it](https://docs.pytest.org/en/latest/reference.html#confval-junit_logging)

miri64 commented:
I don't assume anyone looked at my weekend-shift work, so I just squashed it all together into one :-)

miri64 commented:
This PR is getting quite large. Shall I split it up in one for the framework and one for each spec (with a tracking PR including all release tests implemented so far to ease testing)?

miri64 commented:
> This PR is getting quite large. Shall I split it up in one for the framework and one for each spec (with a tracking PR including all release tests implemented so far to ease testing)?

Squashed in anticipation of that.

miri64 commented:
Ok, I now removed all spec test implementations. It's still big, but there is a lot of ground work done here that is needed by the other PRs.

jia200x commented:
self tests pass and code-wise looks ok overall. I used #164 to run all tests and the base script seems to work.
Since this is a huge improvement in the direction of automated testing, I would be OK to go with it in the current state (we can patch the Ctrl + C issue in a follow up).

Will run again with the last commit and I would say we are ready to go

jia200x commented:
please squash before merging :)

jia200x commented:
(we should maybe add commit lint to this repository too)

miri64 commented:
> (we should maybe add commit lint to this repository too)

Yes, but since there are a lot of legacy scripts still lingering, until everything is integrated, I would prefer to do that after all spec tests are merged (see https://github.com/RIOT-OS/Release-Specs/pull/164/commits/5a95f200d33ab254339382347a847c19596b171f).

miri64 commented:
Squashed.
